Sources/ScipioKit/DescriptionPackage.swift

```diff
@@ -220,43 +245,39 @@ extension DescriptionPackage {
     }
 
     private func buildProducts(from target: ResolvedTarget) throws -> Set<BuildProduct> {
-        guard let package = graph.package(for: target) else {
+        guard let package = descriptionPackage.graph.package(for: target),
+              !visitedTargets.contains(target)
+        else {
             return []
         }
```

Comment:
[Q] If the target was already visited, should it return just an empty set? Don't we need to keep the products resolved?

Reply:
Thank you for your feedback.
The early return of an empty set when a target has already been visited is intentional and should not cause issues in the current implementation.

The return value of `buildProducts()` is used in:

```Swift
Set(try rootTarget.recursiveTargetDependencies().flatMap(buildProducts(from:)))
```
in `resolveBuildProduct()`

```Swift
let productMap: [String: BuildProduct] = Dictionary(products.map { ($0.target.name, $0) }) { $1 }
```
in `resolveBuildProducts()`

In both cases, duplicate `BuildProduct` are eliminated. The early return of an empty set using `visitedTargets` is intended to **preemptively** remove part of the duplicates, so I believe it is not an issue.

Reply:
However, since the output of the `buildProducts(from:)` can change between the first and second execution with the same input, it must be used under the assumption that there is logic to eliminate such duplicates. Therefore, as you suggested, it might be better to keep the products resolved.